### ISSUE TITLE
Oban.Worker 'use' macro can't compile with variables

### ISSUE THREAD

**Issue by ibarakaiev** [adobe/elixir-styler#65]
First of all, thanks for the amazing work on Styler!

## Versions

* Elixir: 1.15.2
* Styler: 0.8.3

## Example Input

As an [example from Oban documentation](https://hexdocs.pm/oban/Oban.Worker.html#module-snoozing-jobs),

```elixir
defmodule MyApp.SnoozingWorker do
  @max_attempts 20

  use Oban.Worker, max_attempts: @max_attempts

  @impl Worker
  def backoff(%Job{} = job) do
    corrected_attempt = @max_attempts - (job.max_attempts - job.attempt)

    Worker.backoff(%{job | attempt: corrected_attempt})
  end

  @impl Worker
  def perform(job) do
    if MyApp.something?(job), do: :ok, else: {:snooze, 60}
  end
end
```

## Stacktrace / Current Behaviour
Currently, the output of Styler is:

```elixir
defmodule MyApp.SnoozingWorker do
  @moduledoc false
  use Oban.Worker, max_attempts: @max_attempts

  @max_attempts 20

  @impl Worker
  def backoff(%Job{} = job) do
    corrected_attempt = @max_attempts - (job.max_attempts - job.attempt)

    Worker.backoff(%{job | attempt: corrected_attempt})
  end

  @impl Worker
  def perform(job) do
    if MyApp.something?(job), do: :ok, else: {:snooze, 60}
  end
end
```

However, this breaks compilation, as `@max_attempts` is inaccessible when `use`ing `Oban.Worker` (it's defined later). The `@moduledoc` example in the troubleshooting guide is similar to this, but the provided solution didn't work for me — and, frankly, this seems like it should be handled differently. In the case of `@moduledoc`, moving module attributes above it would not be acceptable, but with `use` macros it seems it would be a common use case.

**Comment by ibarakaiev** [comment]
Creating a bare Mix project and adding this compiles and works fine:

```elixir
max_attempts = 20

defmodule MyApp.SnoozingWorker do
  @moduledoc "#{max_attempts}"

  use Oban.Worker, queue: :default, max_attempts: 20

  @max_attempts max_attempts

  def run do
    IO.puts(@max_attempts)
  end
end
```

However, using `max_attempts` as a keyword value when `use`ing a macro fails:

```elixir
max_attempts = 20

defmodule MyApp.SnoozingWorker do
  @moduledoc "#{max_attempts}"

  use Oban.Worker, queue: :default, max_attempts: max_attempts

  @max_attempts max_attempts

  def run do
    IO.puts(@max_attempts)
  end
end
```

```
error: undefined variable "max_attempts"
  deps/oban/lib/oban/worker.ex:6: MyApp.SnoozingWorker.__opts__/0


== Compilation error in file lib/worker.ex ==
** (CompileError) lib/worker.ex: cannot compile module MyApp.SnoozingWorker (errors have been logged)
```

**Comment by novaugust** [comment]
hey @ibarakaiev, had a long weekend, sorry for the delay :)

thanks a lot for the repro. something interesting is happening here - i can repro the problem when using exactly your code (oban's use doesn't work), but using a different custom `use` + unquote + after_compile allows things to proceed no problem! 🤔 

here's the code in oban:

https://github.com/sorentwo/oban/blob/43a653f63abda6a287841ea5e62bf44bf7e24d59/lib/oban/worker.ex#L301C22-L311

this more-or-less identical code compiles

```ex
defmodule TestUse do
  def __using__(opts) do
    quote location: :keep do
      @after_compile TestUse
      def __opts__ do
        Keyword.put(unquote(opts), :worker, to_string(__MODULE__))
      end
    end
  end

  defmacro __after_compile__(%{module: module}, _env) do
    Enum.each(module.__opts__(), &IO.inspect/1)
  end
end
```

```ex
a_variable = :woo

defmodule MyTestUse do
  @moduledoc "#{a_variable}"
  use TestUse, a_variable: a_variable
end
```

all of this compiles just fine. so, i'm wondering if something weird happens across dependency boundaries  ¯\\\_(ツ)\_/¯

--------------------------------------------------------------

tldr: huh. i'm going to sit on this for now until i understand what's special about oban's `use` vs `use` in other scenarios. at the least i'll update the title to help others find the bug. sorry for not having a clear solution for you just yet.

**Comment by janpieper** [comment]
Any news on this? I just stumbled upon exactly that issue in my code :confused:

**Comment by novaugust** [comment]
alas, i never got anywhere with it. tried tapping some friends to see if they could figure things out but no one knows what the difference is yet. sorry jan :(

**Comment by novaugust** [comment]
> When `__using__` is defined as a macro then this does fail to compile:

woww what a slipup. cheers juan. now i'm just surprised you can get away with doing either!

anyways, given that's the issue there's really no good workaround for it right now.

**Comment by 03juan** [comment]
Yes I agree. In honouring Styler's intention to strictly enforce Credo's `StrictModuleLayout` my opinion is that Oban is providing technically functional but stylistically incorrect sample code that `ModuleDirectives` specifically warns "**This can break your code**"!

So there no simpler way than to refactor as something like this 🤷

```elixir
max_attempts = 20

defmodule MyApp.SnoozingWorker do
  @moduledoc "#{max_attempts}"

  use Oban.Worker, queue: :default, max_attempts: 20

  @max_attempts max_attempts

  def run do
    IO.puts(@max_attempts)
  end
end
```

**Comment by 03juan** [comment]
Ok, there is another way...
```elixir
defmodule TestUse do
  defmacro __using__(opts) do
    quote location: :keep do
      @after_compile TestUse
      def opts, do: unquote(opts)
    end
  end

  defmacro __after_compile__(%{module: mod}, _env) do
    IO.inspect(mod.opts(), label: "after_compile #{mod}.opts()")
  end
end

a_var = :woo

defmodule A do
  use TestUse, my_var: unquote(a_var)
end

IO.inspect(A.opts(), label: "runtime A.opts()")
```
> after_compile Elixir.A.opts(): [my_var: :woo]
> runtime A.opts(): [my_var: :woo]

😆

**Comment by novaugust** [comment]
eyyy so if folks just do `use ... unquote(my_var)` it works out eh? thanks for the sleuthing juan, i'll get the readme updated and close this issue out =)
